plotter/contour.go

```diff
@@ -0,0 +1,488 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+//go:generate ./list
+
+package plotter
+
+import (
+	"image/color"
+	"math"
+	"sort"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/palette"
+	"github.com/gonum/plot/vg"
+	"github.com/gonum/plot/vg/draw"
+)
+
+// Contour implements the Plotter interface, drawing
+// a contour plot of the values in the GridXYZ field. The
+// row index is from top to bottom.
+type Contour struct {
+	GridXYZ GridXYZ
+
+	// Levels describes the contour heights to plot.
+	Levels []float64
+
+	// LineStyle is the style of the contour lines.
+	draw.LineStyle
+
+	// Palette is the color palette used to render
+	// the heat map.
+	Palette palette.Palette
+
+	// Underflow and Overflow are colors used to draw
+	// contours outside the dynamic range defined
+	// by Min and Max.
+	Underflow color.Color
+	Overflow  color.Color
+
+	// Min and Max define the dynamic range of the
+	// heat map.
+	Min, Max float64
+}
+
+// NewContour creates as new contour plotter for the given data,
+// using the provided palette.
+func NewContour(g GridXYZ, levels []float64, p palette.Palette) *Contour {
+	var min, max float64
+	type minMaxer interface {
+		Min() float64
+		Max() float64
+	}
+	switch g := g.(type) {
+	case minMaxer:
+		min, max = g.Min(), g.Max()
+	default:
+		min, max = math.Inf(1), math.Inf(-1)
+		c, r := g.Dims()
+		for i := 0; i < c; i++ {
+			for j := 0; j < r; j++ {
+				v := g.Z(i, j)
+				if math.IsNaN(v) {
+					continue
+				}
+				min = math.Min(min, v)
+				max = math.Max(max, v)
+			}
+		}
+	}
+
+	if len(levels) == 0 {
+		levels = quantilesR7(g, defaultQuantiles)
+	}
+
+	return &Contour{
+		GridXYZ:   g,
+		Levels:    levels,
+		LineStyle: DefaultLineStyle,
+		Palette:   p,
+		Min:       min,
+		Max:       max,
+	}
+}
+
+// Default quantiles for case where levels is not explicitly set.
+var defaultQuantiles = []float64{0.01, 0.05, 0.25, 0.5, 0.75, 0.95, 0.99}
+
+// quantilesR7 returns the pth quantiles of the data in g according the th R-7 method.
```

Comment:
My opinion is that we should be using gonum/stat here, and if gonum/stat isn't currently sufficient we should improve it. Certainly that would imply coming up with a public name for R7, and possibly implementing a Quantiles function that efficiently finds a set of sorted quantiles instead of just one. Thoughts?